lib/manager/flux/extract.ts

```diff
@@ -0,0 +1,117 @@
+import { loadAll } from 'js-yaml';
+import { HelmDatasource } from '../../datasource/helm';
+import { logger } from '../../logger';
+import { SkipReason } from '../../types';
+import { readLocalFile } from '../../util/fs';
+import type { ExtractConfig, PackageDependency, PackageFile } from '../types';
+import type {
+  FluxManifest,
+  FluxResource,
+  HelmRelease,
+  HelmRepository,
+} from './types';
+
+function readManifest(content: string): FluxManifest | null {
+  const manifest: FluxManifest = { releases: [], repositories: [] };
+  let resources: FluxResource[];
+  try {
+    resources = loadAll(content, null, { json: true }) as FluxResource[];
+  } catch (err) {
+    logger.debug({ err }, 'Failed to parse Flux manifest');
+    return null;
+  }
+
+  // It's possible there are other non-Flux HelmRelease/HelmRepository CRs out there, so we filter based on apiVersion.
+  for (const resource of resources) {
+    switch (resource?.kind) {
+      case 'HelmRelease':
+        if (
+          resource.apiVersion?.startsWith('helm.toolkit.fluxcd.io/') &&
+          resource.spec?.chart?.spec?.chart
+        ) {
+          manifest.releases.push(resource);
+        }
+        break;
+      case 'HelmRepository':
+        if (
+          resource.apiVersion?.startsWith('source.toolkit.fluxcd.io/') &&
+          resource.metadata?.name &&
+          resource.metadata.namespace &&
+          resource.spec?.url
+        ) {
+          manifest.repositories.push(resource);
+        }
+        break;
+    }
+  }
+
+  return manifest;
+}
+
+function resolveReleases(
+  releases: HelmRelease[],
+  repositories: HelmRepository[]
+): PackageDependency[] {
+  return releases.map((release) => {
+    const res: PackageDependency = {
+      depName: release.spec.chart.spec.chart,
+      currentValue: release.spec.chart.spec.version,
+    };
+
+    const matchingRepositories = repositories.filter(
+      (rep) =>
+        rep.kind === release.spec.chart.spec.sourceRef?.kind &&
+        rep.metadata.name === release.spec.chart.spec.sourceRef.name &&
+        rep.metadata.namespace ===
+          (release.spec.chart.spec.sourceRef.namespace ||
+            release.metadata?.namespace)
+    );
+    if (matchingRepositories.length) {
+      res.registryUrls = matchingRepositories.map((repo) => repo.spec.url);
+    } else {
+      res.skipReason = SkipReason.UnknownRegistry;
+    }
+
+    return res;
+  });
+}
+
+export function extractPackageFile(content: string): PackageFile | null {
+  const manifest = readManifest(content);
+  if (!manifest) {
+    return null;
+  }
+  const deps = resolveReleases(manifest.releases, manifest.repositories);
+  return deps.length ? { deps: deps, datasource: HelmDatasource.id } : null;
+}
+
+export async function extractAllPackageFiles(
```

Comment:
`extractAllPackageFiles` is only necessary if package files need to be correlated amongst each other. Is that the case here? It seems `extractPackageFile` is also defined but not referenced internally

Reply:
`extractPackageFile` is required for auto-replace 😕

Reply:
I'm wondering if it's the only one required?

Reply:
Yes, we need `extractAllPackageFiles` since we have to read all `HelmRelease` and `HelmRepository` manifests to correlate `HelmReleases` resources to `HelmRepository` resources. And yes, @viceice is correct, we need `extractPackageFile` for auto-replace - based on my testing, it is the only function required for auto-replace to work (it's just used to validate that the version replacement was successful).